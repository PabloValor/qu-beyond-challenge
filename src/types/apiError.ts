export interface ApiError { 
    message: string
    reason?: string
    level?: number
    trace?: string
}