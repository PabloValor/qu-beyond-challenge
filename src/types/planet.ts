import Film from "./film"
import People  from "./people"

export interface Planet {
    id?: string
    name: string
    diameter: string
    rotation_period: string
    orbital_period: string
    gravity: string
    population: string
    climate: string
    terrain: string
    surface_water: string
    residents: People[]
    films: Film[]
    url: string
    created: string
    edited: string
}

export interface PlanetResponse { 
    count: number
    next: string
    previous: string
    results: Planet[]
}