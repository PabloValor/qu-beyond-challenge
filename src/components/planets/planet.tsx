import { Planet as TPlanet} from "../../types/planet"
import { Badget } from "../util/badget"
import { FaTemperatureHigh } from 'react-icons/fa'
import { IoMdPeople } from 'react-icons/io'
import { Button } from "../util/button"
import slugify from 'slugify'
import { Link } from "react-router-dom"
import { getRandomImageId } from "../../util"

export const displayPopulationNumber = (population: string) => {

    const number = Number(population)

    if(isNaN(number)) return population

    if(number >= 1000000) {
        return `${number / 1000000}M`
    } else if(number >= 1000) { 
        return `${number / 1000}K`
    }
    return number
}


export default function Planet({ planet } : { planet : TPlanet }) { 


    return (
        <div className={`
            p-3
            rounded-md shadow-lg
            cursor-pointer
            w-full md:max-w-[300px]
            bg-[url('/img/bg-galaxy.png')] bg-center bg-cover bg-no-repeat
            relative
            hover:bottom-1
        `}>
            <h2 className="font-bold text-center text-slate-50">{planet.name.toUpperCase()}</h2>

            <Image planet={planet}/>

            <Badget styles="mr-1">
                <FaTemperatureHigh className="inline-block mr-1"/>
                <span data-testid="climate">{planet.climate}</span>
            </Badget>
            <Badget styles="mr-1">
                <IoMdPeople className="inline-block mr-1"/>
                <span data-testid="population" title={`population: ${planet.population}`}>{displayPopulationNumber(planet.population)}</span>
            </Badget>
            
        </div>
    )

    function Image({ planet } : {planet : TPlanet}) { 
        return (
            <Link to={`/planet/${slugify(planet.name, { lower: true })}/${planet.id}`}>
                <div className="relative">
                    <img src={`/img/planet-${getRandomImageId()}.svg`}
                    alt={`Image of planet ${planet.name}`}
                    title={`Image of planet ${planet.name}`}
                    className={`
                        block my-4 mx-auto w-[200px]
                    `}
                    />
                    <div className={`
                        absolute top-0 bottom-0 left-0 right-0
                        flex align-middle items-center justify-around	
                        opacity-0
                        hover:opacity-100 
                        `
                    }>
                        <Button>
                                {`show more`.toUpperCase()}
                        </Button>
                    </div>
                </div>
            </Link>
        )
    }
}
