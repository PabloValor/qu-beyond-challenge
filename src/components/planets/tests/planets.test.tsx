import { ReactTestRenderer, ReactTestRendererJSON, create } from 'react-test-renderer'
import Planets from '../planets'
import { Planet as TPlanet, PlanetResponse } from '../../../types/planet'
import {render, screen} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Planet from '../planet';

// function toJson(component: ReactTestRenderer) {
//     const result = component.toJSON()
//     expect(result).toBeDefined()
//     expect(result).toBeInstanceOf(Array)
//     return result as ReactTestRendererJSON
// }

// describe('Planets component test', () => {

//   // afterEach(() => {
//   //   vi.restoreAllMocks()
//   // })

//   it('shoud match snapshot', () => {
//     const component = create(
//       <Planets/>,
//     )
//     let json = toJson(component)
//     expect(json).toMatchSnapshot()
//   })
  
//   it('should render a planet', () => { 

//     const planetsMock: Planet[] = [
//       {
//         id: "123",
//         name: "Earth",
//         diameter: "12742",
//         rotation_period: "24",
//         orbital_period: "365",
//         gravity: "9.8",
//         population: "7800000000",
//         climate: "Temperate",
//         terrain: "Land, Water",
//         surface_water: "70",
//         residents: [],
//         films: [],
//         url: "https://example.com/earth",
//         created: "2021-01-01",
//         edited: "2021-01-02",
//       }
//     ]

//     const responseApiMock: PlanetResponse = { 
//       count: 1,
//       next: "",
//       previous: "",
//       results: planetsMock
//     }

//     const mocks = vi.hoisted(() => {
//       return { 
//         mockreponse = { 
//           data: responseApiMock,
//           error: undefined,
//           isLoading: false
//         }
//       }
//     })

//     vi.mock('../../../services/swapi', () => { 
//       return { 
//         getPlanets: () => mockreponse
//       }
//     })

//     render(<Planets/>)

//     expect(screen.getByText('/Earth/i')).toBeDefined()

//   })

// })


const planetMock: TPlanet = {
    id: "123",
    name: "Earth",
    diameter: "12742",
    rotation_period: "24",
    orbital_period: "365",
    gravity: "9.8",
    population: "7800000000",
    climate: "Temperate",
    terrain: "Land, Water",
    surface_water: "70",
    residents: [],
    films: [],
    url: "https://example.com/earth",
    created: "2021-01-01",
    edited: "2021-01-02",
}

function toJson(component: ReactTestRenderer) {
const result = component.toJSON()
expect(result).toBeDefined()
return result as ReactTestRendererJSON
}

describe('<Planets />', () => {

test('should match snapshot', () => { 
    const component = create(
    <MemoryRouter>
        <Planets/>
    </MemoryRouter>
    )
    let json = toJson(component)
    expect(json).toMatchSnapshot() 
})

test('Planets mounts properly', () => {
    const wrapper = render(<MemoryRouter> <Planets/></MemoryRouter> )
    expect(wrapper).toBeTruthy()

})
});