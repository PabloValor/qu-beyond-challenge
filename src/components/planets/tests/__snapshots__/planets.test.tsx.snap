// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Planets /> > should match snapshot 1`] = `
[
  <div
    className="flex flex-wrap place-content-center gap-8"
  >
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="
        p-3
        rounded-md shadow-lg
        w-full md:max-w-[300px]
        flex min-h-[22.5rem] h-full
        bg-slate-500
        animate-pulse 
        "
    >
      <div
        className="flex flex-col space-y-3 justify-between items-center w-full animate-pulse"
      >
        <div
          className="h-8 w-full rounded-md bg-slate-400"
        />
        <div
          className="h-[230px] w-[230px] rounded-full bg-slate-400"
        />
        <div
          className="w-full"
        >
          <div
            className="flex flex-row space-x-1"
          >
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
            <div
              className="h-8 w-2/6 rounded-md bg-slate-400"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="flex flex-wrap place-content-center gap-8"
  />,
]
`;
